File: .config/nvim/deinrc/hooks/ddu.lua
-- lua_add {{{
local function start(opts)
  local width = math.floor(vim.o.columns * 0.8)
  local previewWidth = math.floor(width * 0.5) - 1
  local height = math.floor(vim.o.lines * 0.8)
  local previewHeight = height - 2
  local row = math.floor((vim.o.lines - height) / 2)
  local previewRow = row + 1
  local col = math.floor(vim.o.columns * 0.1)
  local halfWidth = math.floor(vim.o.columns * 0.5)
  local previewCol = halfWidth - 1
  local default = {
    uiParams = {
      ff = {
        winWidth = width,
        winHeight = height,
        winRow = row,
        winCol = col,
        previewWidth = previewWidth,
        previewHeight = previewHeight,
        previewRow = previewRow,
        previewCol = previewCol,
      },
    },
  }

  local config = vim.tbl_deep_extend("keep", default, opts)
  vim.fn["ddu#start"](config)
end

local function set_keymaps()
  vim.keymap.set("n", "<leader>d", function() start({ name = "source" }) end)
  vim.keymap.set("i", "<C-x><C-l>", function() start({ name = "copilot" }) end)
  vim.keymap.set(
    "i",
    "<C-x><C-e>",
    function()
      start({
        sources = { { name = "emoji", options = { defaultAction = "append" } } },
        uiParams = { ff = { replaceCol = vim.fn.col(".") } },
      })
    end
  )

  vim.api.nvim_create_autocmd("FileType", {
    pattern = "ddu-ff",
    callback = function()
      local action = vim.fn["ddu#ui#ff#do_action"]
      local opts = { buffer = true }
      vim.keymap.set("n", "<CR>", function() action("itemAction") end, opts)
      vim.keymap.set("n", "<Space>", function() action("toggleSelectItem") end, opts)
      vim.keymap.set("n", "i", function() action("openFilterWindow") end, opts)
      vim.keymap.set("n", "q", function() action("quit") end, opts)
    end,
  })

  vim.api.nvim_create_autocmd("FileType", {
    pattern = "ddu-ff-filter",
    callback = function()
      local opts = { buffer = true }
      vim.keymap.set("i", "<CR>", "<Esc><Cmd>call ddu#ui#do_action('closeFilterWindow')<CR>", opts)
      vim.keymap.set("n", "<CR>", "<Cmd>call ddu#ui#do_action('closeFilterWindow')<CR>", opts)
    end,
  })

  vim.api.nvim_create_autocmd("LspAttach", {
    group = vim.api.nvim_create_augroup("LspAttach_ddu", {}),
    callback = function(ev)
      local opts = { buffer = ev.buf }
      vim.keymap.set("n", "gd", function() start({ name = "lsp-definition" }) end, opts)
      vim.keymap.set("n", "gr", function() start({ name = "lsp-references" }) end, opts)
      vim.keymap.set("n", "<space>q", function() start({ name = "lsp-diagnostic" }) end, opts)
      vim.keymap.set({ "n", "v" }, "<space>ca", function() start({ name = "lsp-codeAction" }) end, opts)
    end,
  })
end

set_keymaps()
-- }}}

-- lua_source {{{
vim.fn["ddu#custom#load_config"](vim.env.DEIN_CONFIG_BASE .. "hooks/ddu.ts")

File: .config/nvim/deinrc/hooks/ddu.ts
import {
  BaseConfig,
  ContextBuilder,
} from "https://deno.land/x/ddu_vim@v3.4.3/types.ts";
import type { Denops } from "https://deno.land/x/ddu_vim@v3.4.3/deps.ts";
import type { Params as FfParams } from "https://deno.land/x/ddu_ui_ff@v1.1.0/ff.ts";

const ffWindowSizeParams = {
  winWidth: "&columns * 4 / 5",
  winHeight: "&lines * 4 / 5",
  winRow: "&lines / 10",
  winCol: "&columns / 10",
  previewWidth: "&columns * 2 / 5 - 1",
  previewHeight: "&lines * 4 / 5 - 2",
  previewRow: "&lines / 10 + 1",
  previewCol: "&columns / 2 - 1",
} as Partial<FfParams>;

export class Config extends BaseConfig {
  // deno-lint-ignore require-await
  override async config(args: {
    denops: Denops;
    contextBuilder: ContextBuilder;
  }): Promise<void> {
    args.contextBuilder.patchLocal("lsp-definition", {
      sync: true,
      sources: [
        {
          name: "lsp_definition",
          params: { method: "textDocument/definition" },
        },
        {
          name: "lsp_definition",
          params: { method: "textDocument/typeDefinition" },
        },
        {
          name: "lsp_definition",
          params: { method: "textDocument/declaration" },
        },
        {
          name: "lsp_definition",
          params: { method: "textDocument/implementation" },
        },
      ],
      uiParams: {
        ff: {
          immediateAction: "open",
        } as Partial<FfParams>,
      },
    });
    args.contextBuilder.patchLocal("lsp-references", {
      sync: true,
      sources: [{ name: "lsp_references" }],
      uiParams: {
        ff: {
          immediateAction: "open",
        } as Partial<FfParams>,
      },
    });
    args.contextBuilder.patchLocal("lsp-diagnostic", {
      sync: true,
      sources: [{
        name: "lsp_diagnostic",
        params: {
          buffer: 0,
        },
      }],
    });
    args.contextBuilder.patchLocal("lsp-codeAction", {
      sync: true,
      sources: [{ name: "lsp_codeAction" }],
    });
    args.contextBuilder.patchLocal("git-ls-files", {
      sources: [{
        name: "file_external",
        params: { "cmd": ["git", "ls-files"] },
      }],
    });
    args.contextBuilder.patchGlobal({
      ui: "ff",
      uiParams: {
        ff: {
          ...ffWindowSizeParams,
          split: "floating",
          floatingBorder: "single",
          filterSplitDirection: "floating",
          previewFloating: true,
          previewFloatingBorder: "single",
          previewSplit: "vertical",
          highlights: {
            floating: "Normal",
            floatingBorder: "Special",
          },
          startAutoAction: true,
          autoAction: {
            name: "preview",
          },
        } as Partial<FfParams>,
      },

      sources: [],
      sourceParams: {},
      sourceOptions: {
        _: {
          ignoreCase: true,
          matchers: ["matcher_substring"],
        },
        copilot: {
          defaultAction: "append",
        },
        emoji: {
          defaultAction: "append",
        },
        line: {
          matchers: ["matcher_kensaku"],
        },
      },

      kindOptions: {
        command_history: { defaultAction: "edit" },
        file: { defaultAction: "open" },
        source: { defaultAction: "execute" },
        ui_select: { defaultAction: "select" },
        url: { defaultAction: "browse" },
        word: { defaultAction: "append" },
        lsp: { defaultAction: "open" },
        lsp_codeAction: { defaultAction: "apply" },
        help: { defaultAction: "tabopen" },
        window: { defaultAction: "open" },
      },
    });
  }
}
